src/devices/timer-src.js

```diff
@@ -0,0 +1,199 @@
+ * key-value pairs with keys (deadlines) less than or equal to some value.
+ *
+ * To support quiescent solo vats (which normally only run if there's an
+ * incoming event), we'd want to tell the host loop when we should next be
+ * scheduled. It might be cheaper to find the smallest index in a sorted map.
+ */
+function buildTimerMap() {
+  const numberToList = new Map();
+  function add(time, cb, r = undefined) {
+    const cbRecord = r ? { cb, r } : { cb };
+    if (!numberToList.has(time)) {
+      numberToList.set(time, [cbRecord]);
+    } else {
+      numberToList.get(time).push(cbRecord);
+    }
+  }
+
+  // We don't expect this to be called often, so we don't optimize for it.
+  function remove(cb) {
+    for (const [time, cbs] of numberToList) {
+      if (cbs.length === 1) {
+        if (cbs[0].cb === cb) {
+          numberToList.delete(time);
+          return cb;
+        }
+      } else {
+        cbs.splice(cbs.indexOf(cb), 1);
+        return cb;
+      }
+    }
+    return null;
+  }
+
+  // Remove and return all pairs indexed by numbers up to target
+  function removeEventsThrough(target) {
+    const returnValues = new Map();
+    for (const [time, cbs] of numberToList) {
+      if (time <= target) {
+        returnValues.set(time, cbs);
+        numberToList.delete(time);
+      }
+    }
+    return returnValues;
+  }
+  return harden({ add, remove, removeEventsThrough });
+}
+
+// curryPollFn provided at top level so it can be exported and tested.
+function curryPollFn(SO, deadlines) {
+  // Now might be Date.now(), or it might be a block height.
+  function poll(now) {
+    const timeAndEvents = deadlines.removeEventsThrough(now);
+    let wokeAnything = false;
+    timeAndEvents.forEach(events => {
+      for (const event of events) {
+        try {
+          if (event.r) {
+            SO(event.cb).wake(event.r);
+          } else {
+            SO(event.cb).wake();
+          }
+          wokeAnything = true;
+        } catch (e) {
+          if (event.r) {
+            event.r.disable();
+          }
+          // continue to wake other events.
+        }
+      }
+    });
+    return harden(wokeAnything);
+  }
+
+  return poll;
+}
+
+// bind the repeater builder over deadlines so it can be exported and tested.
+function curryRepeaterBuilder(deadlines, getLastPolled) {
+  // An object whose presence can be shared with Vat code to enable reliable
+  // repeating schedules. There's no guarantee that the callback will happen at
+  // the precise time, but the repeated calls will reliably be triggered at
+  // consistent intervals.
+  //
+  // The timer can also create Repeater objects that allow holders to schedule
+  // events at regular intervals even though individual callbacks can be
+  // arbitrarily delayed. The Repeaters have access to their startTime and
+  // interval as well as the latest time we were polled. This allows them to
+  // reschedule.
+  function buildRecurringTimeout(startTime, interval) {
```

Comment:
Any chance we could rename this to `buildRepeater`, or change line 157 to `buildRecurringTimeout`, so these two match? It tooks me a while to realize they were the same thing.